src/main/java/graphql/schema/usage/SchemaUsage.java

```diff
@@ -0,0 +1,245 @@
+        }
+        pathSoFar.add(elementName);
+
+        if (ScalarInfo.isGraphqlSpecifiedScalar(elementName)) {
+            return true;
+        }
+
+        List<GraphQLDirective> directives = schema.getDirectives(elementName);
+        if (!directives.isEmpty()) {
+            String directiveName = directives.get(0).getName();
+            if (Directives.isGraphqlSpecifiedDirective(directiveName)) {
+                return true;
+            }
+            if (isNamedElementReferenced(schema, directiveName, missCache, pathSoFar)) {
+                return true;
+            }
+        }
+
+        GraphQLNamedType type = schema.getTypeAs(elementName);
+        if (type == null) {
+            return false;
+        }
+        if (missCache.contains(elementName)) {
+            return false;
+        }
+        if (Introspection.isIntrospectionTypes(type)) {
+            return true;
+        }
+
+        if (type == schema.getQueryType()) {
+            return true;
+        }
+        if (type == schema.getMutationType()) {
+            return true;
+        }
+        if (type == schema.getSubscriptionType()) {
+            return true;
+        }
+
+        if (isNamedElementReferenced(schema, elementName, missCache, pathSoFar)) {
+            return true;
+        }
+
+        if (type instanceof GraphQLInterfaceType) {
+            Set<String> implementors = builder.interfaceImplementors.getOrDefault(type.getName(), emptySet());
+            for (String implementor : implementors) {
+                if (isReferencedImpl(schema, implementor, missCache, pathSoFar)) {
+                    return true;
+                }
+            }
+        }
+        if (type instanceof GraphQLObjectType) {
+            List<GraphQLNamedOutputType> interfaces = ((GraphQLObjectType) type).getInterfaces();
+            for (GraphQLNamedOutputType memberInterface : interfaces) {
+                Set<String> implementors = builder.interfaceImplementors.getOrDefault(memberInterface.getName(), emptySet());
+                for (String implementor : implementors) {
+                    if (isReferencedImpl(schema, implementor, missCache, pathSoFar)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean isNamedElementReferenced(GraphQLSchema schema, String elementName, Set<String> missCache, Set<String> pathSoFar) {
+        Set<String> references = builder.typeBackReferences.getOrDefault(elementName, emptySet());
+        for (String reference : references) {
+            if (isReferencedImpl(schema, reference, missCache, pathSoFar)) {
+                return true;
+            }
+        }
+        missCache.add(elementName);
+        return false;
+    }
+
+    @Internal
+    static class Builder {
+        Map<String, Integer> fieldReferenceCounts = new HashMap<>();
+        Map<String, Integer> inputFieldReferenceCounts = new HashMap<>();
```

Comment:
I would recommend ImmutableMaps.Builder here, which preserve the order and are immutable OR our beloved `LinkedHashMap` or a comment why the order is never interesting.

Reply:
ImmutableMap.Builder does not support `compute` like behavior.  I want to accumulate values (counts) against type names.

```
            private BiFunction<String, Integer, Integer> incCount() {
                return (k, v) -> v == null ? 1 : v + 1;
            }

builder.fieldReferenceCounts.compute(fieldType.getName(), incCount());
```

With map builder it has only PUT and no read capability for past values

So I cant use it.  I do use it on getters however.  I will improve it to create these once and once only